src/Data/LLVM/BitCode/GetBits.hs

```diff
@@ -11,68 +11,106 @@ module Data.LLVM.BitCode.GetBits (
   , skip
   ) where
 
-import Data.LLVM.BitCode.BitString
+import           Data.LLVM.BitCode.BitString
 
-import Control.Applicative (Alternative(..))
-import Control.Arrow (first)
-import Control.Monad (MonadPlus(..),when,replicateM_)
-import qualified Data.Binary.Get as BG
-import Data.Bits (shiftR)
-import Data.ByteString (ByteString)
-import Data.Word (Word32)
+import           Control.Applicative ( Alternative(..) )
+import           Control.Monad ( MonadPlus(..) )
+import           Data.Bits ( (.&.), (.|.), shiftL, shiftR )
+import           Data.ByteString ( ByteString )
+import qualified Data.ByteString as BS
 
 #if !MIN_VERSION_base(4,13,0)
-import Control.Monad.Fail( MonadFail )
+import           Control.Monad.Fail ( MonadFail )
 import qualified Control.Monad.Fail
 #endif
 
 -- Bit-level Parsing -----------------------------------------------------------
 
-newtype GetBits a = GetBits { unGetBits :: SubWord -> BG.Get (a,SubWord) }
+newtype GetBits a = GetBits { unGetBits :: Either String (BitPosition -> BS.ByteString -> (BitsGetter a, BitPosition)) } -- Left is outer fail
+
+newtype BitPosition = BitPosition (NumBits, NumBits) -- (current, limit)
```

Comment:
I think I would prefer `BitPosition` to be a data type with two fields rather than a newtype around a pair, as the latter is slightly more indirect.

Reply:
The base purpose of effort I'm undertaking here is to improve performance and memory utilization, so while in general I am in agreement with avoiding the use of "bare" tuples,  I've been trying to eliminate constructors (yes, this will currently have a tuple constructor, aka `,`, but in a future PR this will change to an unlifted type alias (`type BitPosition = (# ... #)`).  Are you OK with deferring your preference for the moment given this context?

Reply:
Ah, I didn't realize that this was fated to become an unboxed tuple. In that case, I'm fine with this approach.